feat(guide): port and update lots of stuff (#9385)
discordjs/discord.js: apps/guide/src/content/02-creating-your-bot/03-command-handling.mdx

<Alert title="Tip" type="success">
For fully functional slash commands, there are three important pieces of code that need to be written. They are:

1. The [individual command files](slash-commands), containing their definitions and functionality.
2. The command handler, which dynamically reads the files and executes the commands.
3. The [command deployment script](command-deployment), to register your slash commands with Discord so they appear in the interface.

These steps can be done in any order, but **all are required** before the commands are fully functional.

This page details how to complete **Step 2**. Make sure to also complete the other pages linked above!

</Alert>

## Loading command files

Now that your command files have been created, your bot needs to load these files on startup.

In your _`index.js`_ file, make these additions to the base template:

<CH.Code>

```js index.js focus=1:2,8
import { readdirSync } from 'node:fs';
import { join } from 'node:path';

> // focus[18:27]

Reviewer: Technically, the addition of the comma is also important:

```suggestion
// focus[18:28]
```

import { Client, Collection, Events, GatewayIntentBits } from 'discord.js';
import config from './config.json';

> const client = new Client({ intents: [GatewayIntentBits.Guilds] });

Reviewer: https://github.com/discordjs/discord.js/blob/0d0e4d1cb61e27d476352294a5624fec2bbf657b/apps/guide/src/content/02-creating-your-bot/01-initial-files.mdx?plain=1#L201

Should probably put that inside an array or use `|`?

Author: Yeah but that's a different file? Kinda out of scope of this PR

client.commands = new Collection();
```

</CH.Code>

> We recommend attaching a _`.commands`_ property to your client instance so that you can access your commands in other files. The rest of the examples in this guide will follow this convention. For TypeScript users, we recommend extending the base Client class to add this property, [casting](https://www.typescripttutorial.net/typescript-tutorial/type-casting/), or [augmenting the module type](https://www.typescriptlang.org/docs/handbook/modules.html#ambient-modules).

Reviewer: Could we just show this in the above code block? Add a tab, so it shows the module augmentation then what we already have?

Author: This kinda sucks to do because we then need to define a type for the command module.

`{ data: SlashCommandBuilder, execute: async (interaction): void => {} }` isn't really correct, particularly because of the way Builders narrow their type. `Collection<string, any>`?

Reviewer: https://canary.discord.com/channels/222078108977594368/1078692758153351278/1097640194666139698

<Alert title="Tip" type="success">
	- The [_`fs`_](https://nodejs.org/api/fs.html) module is Node's native file system module. _`readdirSync`_ is used to
	read the _`commands`_ directory and identify our command files. - The [_`path`_](https://nodejs.org/api/path.html)
	module is Node's native path utility module. _`join`_ helps construct paths to access files and directories. One of
	the advantages of _`path.join`_ is that it automatically detects the operating system and uses the appropriate

> 	joiners. - The <DocsLink type="class" parent="Collection" /> class extends JavaScript's native
> 	[_`Map`_](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) class, and includes
> 	more extensive, useful functionality. _`Collection`_ is used to store and efficiently retrieve commands for execution.

Reviewer 2: Is this supposed to be on a new line to have 2 bullet points here instead?

```suggestion
	joiners.
	- The <DocsLink type="class" parent="Collection" /> class extends JavaScript's native
	[_`Map`_](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) class, and includes
	more extensive, useful functionality. _`Collection`_ is used to store and efficiently retrieve commands for execution.
```

Reviewer: ```diff
- <DocsLink type="class" parent="Collection" />
+ <DocsLink package="Collection" type="Class" parent="Collection" />
```

</Alert>

Next, using the modules imported above, dynamically retrieve your command files with a few more additions to the _`index.js`_ file:

<CH.Code>

```js index.js mark=3:15 focus=3:15
client.commands = new Collection();

> const commandsPath = join(__dirname, 'commands');

Reviewer: `__dirname` is not defined in ESM.

Reviewer 3: Alternatively, this can be used instead:
```suggestion
const commandsPath = new URL('./commands', import.meta.url);
```

Although this will require either adapting the rest of the code to use [`URL`](https://nodejs.org/dist/latest-v19.x/docs/api/url.html#new-urlinput-base)'s API, or use [`fileURLToPath`](https://nodejs.org/dist/latest-v19.x/docs/api/url.html#urlfileurltopathurl) without further modifications.

> const commandFiles = readdirSync(commandsPath).filter((file) => file.endsWith('.js'));

Reviewer 3: If we are using ESM, shouldn't we take advantage of [top level await](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await#top_level_await)? Or is this intentionally done to be closer to the CJS counterpart?

Author: Happy to go either way - is there any benefit to awaiting readdir from fs/promises over this?

Reviewer 2: It's just more semantic code.

for (const file of commandFiles) {
	const filePath = join(commandsPath, file);

> 	const command = require(filePath);

Reviewer 3: You cannot require ESM modules, `import` is required.

With the correct files identified, the last step is to loop over the array and dynamically set each command into the _`client.commands`_ Collection. For each file being loaded, check that it has at least the _`data`_ and _`execute`_ properties. This helps to prevent errors resulting from loading empty, unfinished or otherwise incorrect command files while you're still developing.

## Receiving command interactions

Every slash command is an _`interaction`_, so to respond to a command, you need to create a listener for the <DocsLink type="class" parent="Client" symbol="e-interactionCreate" /> event that will execute code when your application receives an interaction. Place the code below in the _`index.js`_ file you created earlier.

<CH.Code>

```js index.js
client.on(Events.InteractionCreate, (interaction) => {
	console.log(interaction);
});
```

</CH.Code>

> Not every interaction is a slash command (e.g. _`MessageComponent`_ interactions). Make sure to only handle slash commands in this function by making use of the <DocsLink type="class" parent="BaseInteraction" symbol="isChatInputCommand" /> method to exit the handler if another type is encountered. This method also provides typeguarding for TypeScript users, narrowing the type from _`BaseInteraction`_ to <DocsLink type="class" parent="ChatInputCommandInteraction" />.

Reviewer: ```suggestion
Not every interaction is a slash command (e.g. _`MessageComponent`_ interactions). Make sure to only handle slash commands in this function by making use of the <DocsLink type="class" parent="BaseInteraction" symbol="isChatInputCommand" brackets /> method to exit the handler if another type is encountered. This method also provides type guarding for TypeScript users, narrowing the type from _`BaseInteraction`_ to <DocsLink type="class" parent="ChatInputCommandInteraction" />.
```

	const command = interaction.client.commands.get(interaction.commandName);

	if (!command) {
		console.error(`No command matching ${interaction.commandName} was found.`);
		return;
	}

	try {
		await command.execute(interaction);
	} catch (error) {
		console.error(error);
		if (interaction.replied || interaction.deferred) {
			await interaction.followUp({ content: 'There was an error while executing this command!', ephemeral: true });
		} else {
			await interaction.reply({ content: 'There was an error while executing this command!', ephemeral: true });
		}
	}
});
```

</CH.Code>

## Command categories

So far, all of your command files are in a single _`commands`_ folder. This is fine at first, but as your project grows, the number of files in the _`commands`_ folder will too. Keeping track of that many files can be a little tough. To make this a little easier, you can categorize your commands and put them in subfolders inside the _`commands`_ folder. You will have to make a few changes to your existing code in _`index.js`_ for this to work out.

If you've been following along, your project structure should look something like this:

> ![Project structure before sorting](./images/before-sorting.png)

Reviewer: Right now, I've put all guide-related images in /public/assets. If you want to move these over now (or take updated ones), up to you.

After moving your commands into subfolders, it will look something like this:

> ![Project structure after sorting](./images/after-sorting.png)

Reviewer: Right now, I've put all guide-related images in /public/assets. If you want to move these over now (or take updated ones), up to you.

<Alert title="Warning" type="success">
	Make sure you put every command file you have inside one of the new subfolders. Leaving a command file directly under
	the _`commands`_ folder will create problems.
</Alert>

It is not necessary to name your subfolders exactly like we have named them here. You can create any number of subfolders and name them whatever you want. Although, it is a good practice to name them according to the type of commands stored inside them.

Back in your _`index.js`_ file, where the code to [dynamically read command files](#loading-command-files) is, use the same pattern to read the subfolder directories, and then require each command inside them.

<CH.Code>

```js index.js mark=2:3,5:7,17
client.commands = new Collection();
const foldersPath = join(__dirname, 'commands');
const commandFolders = readdirSync(foldersPath);

for (const folder of commandFolders) {
	const commandsPath = join(foldersPath, folder);
	const commandFiles = readdirSync(commandsPath).filter((file) => file.endsWith('.js'));
	for (const file of commandFiles) {
		const filePath = join(commandsPath, file);

> 		const command = require(filePath);

Reviewer: This probably needs an `await import()`.
